File: go/jacs_document_store_ffi_test.go
package haiai

// Issue 025 — Go FFI tests for the 7 D5/D9 JACS Document Store methods.
//
// Exercises every D5 (SaveMemory / SaveSoul / GetMemory / GetSoul) and D9
// (StoreTextFile / StoreImageFile / GetRecordBytes) wrapper through the
// FFIClient interface using the mockFFIClient test infrastructure. The
// fixture file `fixtures/ffi_method_parity.json` declares these methods in
// the `jacs_document_store` group; this test file pins their wire shape
// (argument names, return types, error mapping) at the adapter boundary.
//
// Mock-only: these tests do NOT load libhaiigo. The full HTTP round-trip is
// exercised by `haisdk/rust/haiai/tests/jacs_remote_integration.rs`
// (`--ignored` against a live hosted stack).

import (
	"bytes"
	"encoding/json"
	"testing"
)

// =============================================================================
// D5 — MEMORY / SOUL wrappers
// =============================================================================

func TestSaveMemoryCapturesContent(t *testing.T) {
	mock := newMockFFIClient("http://localhost:0", "agent-test", "")
	var captured string
	mock.saveMemoryFn = func(content string) (string, error) {
		captured = content
		return "mem-id:v1", nil
	}
	key, err := mock.SaveMemory("# MEMORY.md\n\nproject: foo")
	if err != nil {
		t.Fatalf("SaveMemory returned error: %v", err)
	}
	if key != "mem-id:v1" {
		t.Errorf("expected key=mem-id:v1, got %q", key)
	}
	if captured != "# MEMORY.md\n\nproject: foo" {
		t.Errorf("captured content mismatch: got %q", captured)
	}
}

func TestSaveSoulCapturesContent(t *testing.T) {
	mock := newMockFFIClient("http://localhost:0", "agent-test", "")
	var captured string
	mock.saveSoulFn = func(content string) (string, error) {
		captured = content
		return "soul-id:v1", nil
	}
	key, err := mock.SaveSoul("# SOUL.md\n\nvoice: terse")
	if err != nil {
		t.Fatalf("SaveSoul returned error: %v", err)
	}
	if key != "soul-id:v1" {
		t.Errorf("expected key=soul-id:v1, got %q", key)
	}
	if captured != "# SOUL.md\n\nvoice: terse" {
		t.Errorf("captured content mismatch: got %q", captured)
	}
}

func TestGetMemoryReturnsEnvelopeJSON(t *testing.T) {
	mock := newMockFFIClient("http://localhost:0", "agent-test", "")
	envelope := `{"jacsId":"mem-1","jacsType":"memory","body":"x"}`
	mock.getMemoryFn = func() (string, error) {
		return envelope, nil
	}
	out, err := mock.GetMemory()
	if err != nil {
		t.Fatalf("GetMemory returned error: %v", err)
	}
	if out != envelope {
		t.Errorf("envelope mismatch: got %q, want %q", out, envelope)
	}
}

func TestGetSoulReturnsEnvelopeJSON(t *testing.T) {
	mock := newMockFFIClient("http://localhost:0", "agent-test", "")
	envelope := `{"jacsId":"soul-1","jacsType":"soul"}`
	mock.getSoulFn = func() (string, error) {
		return envelope, nil
	}
	out, err := mock.GetSoul()
	if err != nil {
		t.Fatalf("GetSoul returned error: %v", err)
	}
	if out != envelope {
		t.Errorf("envelope mismatch: got %q, want %q", out, envelope)
	}
}

// =============================================================================
// D9 — typed-content helpers
// =============================================================================

func TestStoreTextFileCapturesPath(t *testing.T) {
	mock := newMockFFIClient("http://localhost:0", "agent-test", "")
	var captured string
	mock.storeTextFileFn = func(path string) (string, error) {
		captured = path
		return "txt-id:v1", nil
	}
	key, err := mock.StoreTextFile("/tmp/signed.md")
	if err != nil {
		t.Fatalf("StoreTextFile returned error: %v", err)
	}
	if key != "txt-id:v1" {
		t.Errorf("expected key=txt-id:v1, got %q", key)
	}
	if captured != "/tmp/signed.md" {
		t.Errorf("captured path mismatch: got %q", captured)
	}
}

func TestStoreImageFileCapturesPath(t *testing.T) {
	mock := newMockFFIClient("http://localhost:0", "agent-test", "")
	var captured string
	mock.storeImageFileFn = func(path string) (string, error) {
		captured = path
		return "png-id:v1", nil
	}
	key, err := mock.StoreImageFile("/tmp/signed.png")
	if err != nil {
		t.Fatalf("StoreImageFile returned error: %v", err)
	}
	if key != "png-id:v1" {
		t.Errorf("expected key=png-id:v1, got %q", key)
	}
	if captured != "/tmp/signed.png" {
		t.Errorf("captured path mismatch: got %q", captured)
	}
}

func TestGetRecordBytesReturnsBytes(t *testing.T) {
	mock := newMockFFIClient("http://localhost:0", "agent-test", "")
	pngMagic := []byte{0x89, 'P', 'N', 'G', 0x0D, 0x0A, 0x1A, 0x0A}
	mock.getRecordBytesFn = func(key string) ([]byte, error) {
		if key != "png-id:v1" {
			t.Errorf("unexpected key %q", key)
		}
		return pngMagic, nil
	}
	out, err := mock.GetRecordBytes("png-id:v1")
	if err != nil {
		t.Fatalf("GetRecordBytes returned error: %v", err)
	}
	if !bytes.Equal(out, pngMagic) {
		t.Errorf("byte mismatch: got %v, want %v", out, pngMagic)
	}
}

// =============================================================================
// Generic JACS Document Store CRUD — also part of the 20-method scope
// =============================================================================

func TestSignAndStorePassesDataJSON(t *testing.T) {
	mock := newMockFFIClient("http://localhost:0", "agent-test", "")
	var captured string
	mock.signAndStoreFn = func(dataJSON string) (json.RawMessage, error) {
		captured = dataJSON
		return json.RawMessage(`{"key":"id1:v1"}`), nil
	}
	out, err := mock.SignAndStore(`{"hello":"world"}`)
	if err != nil {
		t.Fatalf("SignAndStore returned error: %v", err)
	}
	if !bytes.Equal(out, []byte(`{"key":"id1:v1"}`)) {
		t.Errorf("output mismatch: got %s", string(out))
	}
	if captured != `{"hello":"world"}` {
		t.Errorf("captured data mismatch: got %q", captured)
	}
}

func TestSearchDocumentsForwardsArgs(t *testing.T) {
	mock := newMockFFIClient("http://localhost:0", "agent-test", "")
	var capturedQuery string
	var capturedLimit, capturedOffset int
	mock.searchDocumentsFn = func(query string, limit, offset int) (json.RawMessage, error) {
		capturedQuery = query
		capturedLimit = limit
		capturedOffset = offset
		return json.RawMessage(`{"results":[],"total_count":0}`), nil
	}
	_, err := mock.SearchDocuments("marker-xyz", 10, 0)
	if err != nil {
		t.Fatalf("SearchDocuments returned error: %v", err)
	}
	if capturedQuery != "marker-xyz" || capturedLimit != 10 || capturedOffset != 0 {
		t.Errorf("args mismatch: q=%q limit=%d offset=%d", capturedQuery, capturedLimit, capturedOffset)
	}
}

func TestQueryByTypeForwardsArgs(t *testing.T) {
	mock := newMockFFIClient("http://localhost:0", "agent-test", "")
	mock.queryByTypeFn = func(docType string, limit, offset int) ([]string, error) {
		if docType != "memory" || limit != 25 || offset != 0 {
			t.Errorf("unexpected args: %s %d %d", docType, limit, offset)
		}
		return []string{}, nil
	}
	_, err := mock.QueryByType("memory", 25, 0)
	if err != nil {
		t.Fatalf("QueryByType returned error: %v", err)
	}
}

// Issue: array-shaped return for the five trait methods. The trait returns
// `Vec<String>` so the Go interface MUST surface `[]string`. Pin this.
func TestListDocumentsReturnsStringSlice(t *testing.T) {
	mock := newMockFFIClient("http://localhost:0", "agent-test", "")
	mock.listDocumentsFn = func(jacsType string) ([]string, error) {
		return []string{"id1:v1", "id2:v1"}, nil
	}
	out, err := mock.ListDocuments("")
	if err != nil {
		t.Fatalf("ListDocuments returned error: %v", err)
	}
	if len(out) != 2 || out[0] != "id1:v1" {
		t.Errorf("expected [id1:v1, id2:v1], got %v", out)
	}
}

func TestRemoveDocumentReturnsNil(t *testing.T) {
	mock := newMockFFIClient("http://localhost:0", "agent-test", "")
	called := false
	mock.removeDocumentFn = func(key string) error {
		if key != "id1:v1" {
			t.Errorf("expected key=id1:v1, got %q", key)
		}
		called = true
		return nil
	}
	err := mock.RemoveDocument("id1:v1")
	if err != nil {
		t.Fatalf("RemoveDocument returned error: %v", err)
	}
	if !called {
		t.Error("RemoveDocument did not invoke stub")
	}
}

// =============================================================================
// FFI surface area — every D5/D9 method appears in the parity fixture.
// =============================================================================

func TestD5MethodsAreInParityFixture(t *testing.T) {
	fixture := loadParityFixture(t)
	expected := []string{"save_memory", "save_soul", "get_memory", "get_soul"}
	all := make(map[string]bool)
	for _, group := range fixture.Methods {
		for _, m := range group {
			all[m.Name] = true
		}
	}
	for _, name := range expected {
		if !all[name] {
			t.Errorf("D5 method %q missing from ffi_method_parity.json", name)
		}
	}
}

func TestD9MethodsAreInParityFixture(t *testing.T) {
	fixture := loadParityFixture(t)
	expected := []string{"store_text_file", "store_image_file", "get_record_bytes"}
	all := make(map[string]bool)
	for _, group := range fixture.Methods {
		for _, m := range group {
			all[m.Name] = true
		}
	}
	for _, name := range expected {
		if !all[name] {
			t.Errorf("D9 method %q missing from ffi_method_parity.json", name)
		}
	}
}

// =============================================================================
// Real `ffi.Client` doc-store wiring
//
// The 20 doc-store methods previously returned a "not yet wired through
// libhaiigo" stub error. Wiring is now complete (TASK_004 of the
// JACS_DOCUMENT_STORE_FFI_PRD). Real-binding tests live in
// `go/ffi_native_smoke_test.go` (`//go:build cgo_smoke`).
// =============================================================================

File: node/tests/jacs-document-store-ffi.test.ts
/**
 * Issue 025 — Node FFI tests for the 7 D5/D9 JACS Document Store methods.
 *
 * Exercises every D5 (saveMemory / saveSoul / getMemory / getSoul) and D9
 * (storeTextFile / storeImageFile / getRecordBytes) wrapper through the
 * FFIClientAdapter using the createMockFFI test infrastructure. The fixture
 * file `fixtures/ffi_method_parity.json` declares these methods in the
 * `jacs_document_store` group; this test file pins their wire shape
 * (argument names, return types, error mapping) at the adapter boundary.
 *
 * Mock-only: these tests do NOT load the haiinpm native library. The full
 * HTTP round-trip is exercised by `haisdk/rust/haiai/tests/jacs_remote_integration.rs`
 * (`--ignored` against a live hosted stack).
 */

import { describe, expect, it } from 'vitest';
import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';
import { createMockFFI } from './ffi-mock.js';

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function loadParityFixture(): {
  methods: Record<string, Array<{ name: string; args: string[]; returns: string }>>;
} {
  const fixturePath = resolve(__dirname, '../../fixtures/ffi_method_parity.json');
  return JSON.parse(readFileSync(fixturePath, 'utf-8'));
}

// ---------------------------------------------------------------------------
// D5 — MEMORY / SOUL wrappers
// ---------------------------------------------------------------------------

describe('D5 MEMORY / SOUL wrappers (Issue 025)', () => {
  it('saveMemory passes content through to native', async () => {
    const ffi = createMockFFI({
      saveMemory: async (content) => {
        expect(content).toBe('# MEMORY.md\n\nproject: foo');
        return 'mem-id:v1';
      },
    });
    const key = await ffi.saveMemory('# MEMORY.md\n\nproject: foo');
    expect(key).toBe('mem-id:v1');
  });

  it('saveMemory accepts null for default-file mode', async () => {
    let captured: string | null | undefined = 'unset';
    const ffi = createMockFFI({
      saveMemory: async (content) => {
        captured = content;
        return 'mem-id:v2';
      },
    });
    const key = await ffi.saveMemory(null);
    expect(key).toBe('mem-id:v2');
    expect(captured).toBeNull();
  });

  it('saveSoul passes content through to native', async () => {
    const ffi = createMockFFI({
      saveSoul: async (content) => {
        expect(content).toBe('# SOUL.md\n\nvoice: terse');
        return 'soul-id:v1';
      },
    });
    const key = await ffi.saveSoul('# SOUL.md\n\nvoice: terse');
    expect(key).toBe('soul-id:v1');
  });

  it('getMemory returns envelope JSON when present', async () => {
    const envelope = JSON.stringify({
      jacsId: 'mem-1',
      jacsType: 'memory',
      body: 'x',
    });
    const ffi = createMockFFI({
      getMemory: async () => envelope,
    });
    const out = await ffi.getMemory();
    expect(out).toBe(envelope);
  });

  it('getMemory returns null when no record exists', async () => {
    const ffi = createMockFFI({
      getMemory: async () => null,
    });
    const out = await ffi.getMemory();
    expect(out).toBeNull();
  });

  it('getSoul returns envelope JSON when present', async () => {
    const envelope = JSON.stringify({ jacsId: 'soul-1', jacsType: 'soul' });
    const ffi = createMockFFI({
      getSoul: async () => envelope,
    });
    const out = await ffi.getSoul();
    expect(out).toBe(envelope);
  });
});

// ---------------------------------------------------------------------------
// D9 — typed-content helpers
// ---------------------------------------------------------------------------

describe('D9 typed-content helpers (Issue 025)', () => {
  it('storeTextFile passes path through to native', async () => {
    const ffi = createMockFFI({
      storeTextFile: async (path) => {
        expect(path).toBe('/tmp/signed.md');
        return 'txt-id:v1';
      },
    });
    const key = await ffi.storeTextFile('/tmp/signed.md');
    expect(key).toBe('txt-id:v1');
  });

  it('storeImageFile passes path through to native', async () => {
    const ffi = createMockFFI({
      storeImageFile: async (path) => {
        expect(path).toBe('/tmp/signed.png');
        return 'png-id:v1';
      },
    });
    const key = await ffi.storeImageFile('/tmp/signed.png');
    expect(key).toBe('png-id:v1');
  });

  it('getRecordBytes returns Uint8Array', async () => {
    const pngMagic = new Uint8Array([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]);
    const ffi = createMockFFI({
      getRecordBytes: async (key) => {
        expect(key).toBe('png-id:v1');
        return pngMagic;
      },
    });
    const out = await ffi.getRecordBytes('png-id:v1');
    expect(out).toBeInstanceOf(Uint8Array);
    expect(out).toEqual(pngMagic);
  });
});

// ---------------------------------------------------------------------------
// Generic JACS Document Store CRUD — also part of the 20-method scope
// ---------------------------------------------------------------------------

describe('Generic JACS Document Store CRUD (Issue 025)', () => {
  it('signAndStore passes data JSON through and returns Record', async () => {
    const ffi = createMockFFI({
      signAndStore: async (dataJson) => {
        expect(dataJson).toBe('{"hello":"world"}');
        return { key: 'id1:v1', json: '{}' };
      },
    });
    const out = await ffi.signAndStore('{"hello":"world"}');
    expect(out).toEqual({ key: 'id1:v1', json: '{}' });
  });

  it('searchDocuments forwards limit + offset', async () => {
    const ffi = createMockFFI({
      searchDocuments: async (query, limit, offset) => {
        expect(query).toBe('marker-xyz');
        expect(limit).toBe(10);
        expect(offset).toBe(0);
        return { results: [], total_count: 0 };
      },
    });
    const out = await ffi.searchDocuments('marker-xyz', 10, 0);
    expect(out).toEqual({ results: [], total_count: 0 });
  });

  it('queryByType forwards three args', async () => {
    const ffi = createMockFFI({
      queryByType: async (docType, limit, offset) => {
        expect(docType).toBe('memory');
        expect(limit).toBe(25);
        expect(offset).toBe(0);
        return { items: [] };
      },
    });
    const out = await ffi.queryByType('memory', 25, 0);
    expect(out).toEqual({ items: [] });
  });

  it('storageCapabilities takes no args', async () => {
    const ffi = createMockFFI({
      storageCapabilities: async () => ({ fulltext: true, vector: false }),
    });
    const out = await ffi.storageCapabilities();
    expect(out).toEqual({ fulltext: true, vector: false });
  });

  it('removeDocument returns void', async () => {
    let called = false;
    const ffi = createMockFFI({
      removeDocument: async (key) => {
        expect(key).toBe('id1:v1');
        called = true;
      },
    });
    const out = await ffi.removeDocument('id1:v1');
    expect(out).toBeUndefined();
    expect(called).toBe(true);
  });
});

// ---------------------------------------------------------------------------
// FFI surface area — every D5/D9 method appears in the parity fixture.
// ---------------------------------------------------------------------------

describe('D5/D9 methods are in parity fixture (Issue 025)', () => {
  const D5_METHODS = ['save_memory', 'save_soul', 'get_memory', 'get_soul'];
  const D9_METHODS = ['store_text_file', 'store_image_file', 'get_record_bytes'];

  function fixtureMethodNames(): Set<string> {
    const fixture = loadParityFixture();
    const names = new Set<string>();
    for (const group of Object.values(fixture.methods)) {
      for (const m of group) {
        names.add(m.name);
      }
    }
    return names;
  }

  it('all D5 methods appear in parity', () => {
    const parity = fixtureMethodNames();
    for (const name of D5_METHODS) {
      expect(parity.has(name), `missing ${name}`).toBe(true);
    }
  });

  it('all D9 methods appear in parity', () => {
    const parity = fixtureMethodNames();
    for (const name of D9_METHODS) {
      expect(parity.has(name), `missing ${name}`).toBe(true);
    }
  });
});
